### PR TITLE
[APIC-305] Require sql_type for every column when table_columns is provided

### PR DESCRIPTION
**Pull request by izzy-zelichenko**
[APIC-305](https://civisanalytics.atlassian.net/browse/APIC-305)

Ensure data types are detected when the destination table does not exist or it is being dropped and table_columns are provided but do not contain a sql_type for every column.

### CODE REVIEW

File: civis/io/_tables.py [hunk 1]
```diff
@@ -1029,8 +1029,16 @@ def civis_file_to_table(file_id, database, table, client=None,
 
     # Use Preprocess endpoint to get the table columns as needed
     # and perform necessary file cleaning
-    need_table_columns = ((not table_exists or existing_table_rows == 'drop')
+    table_columns_have_sql = True
```
Comment (uttercm) [inline civis/io/_tables.py:1032]: nit: this seems a bit wordy maybe something like `table_columns_valid`, idk, thoughts?

File: civis/io/_tables.py [hunk 2]
```diff
@@ -1029,8 +1029,16 @@ def civis_file_to_table(file_id, database, table, client=None,
...
+    if table_columns:
+        for col in table_columns:
+            if not col['sql_type']:
+                table_columns_have_sql = False
+                break
+
+    need_table_columns = (((not table_exists or existing_table_rows == 'drop')
```
Comment (uttercm) [inline civis/io/_tables.py:1039]: I'm now wondering why we shouldn't throw an error when the table_columns are provided but don't have a `sql_type`, seems to me we might want to throw an error rather than ignore the provided ones?

Comment (izzy-zelichenko) [reply civis/io/_tables.py:1039]: @uttercm ahh do you mean because it'll override all of them, even though some of them might be provide?

File: civis/io/_tables.py (hunk 1 above)
Comment (izzy-zelichenko) [reply civis/io/_tables.py:1032]: Yeah, that seems better. I wanted to be specific, but also thought it sounded awkward as is.

File: civis/io/_tables.py (hunk 2 above)
Comment (uttercm) [reply civis/io/_tables.py:1039]: ya exactly, so they might be confused if the sql_types they did provide are different than the actual types we create since not all were provided

Comment (izzy-zelichenko) [reply civis/io/_tables.py:1039]: That's a good point, but that would be a change in approach, so let me touch base with John on that.

Comment (jkulzick) [reply civis/io/_tables.py:1039]: This makes sense. Just to be clear if no sql types are provided we will detect them but if some but not all are provided we will raise an error.

Comment (izzy-zelichenko) [reply civis/io/_tables.py:1039]: Ah okay, I misunderstood the criteria in the first place. Thanks!

File: civis/tests/test_io.py [hunk 3]
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
             **expected_kwargs
         )
 
+    @pytest.mark.civis_file_to_table
+    @mock.patch('civis.io._tables._process_cleaning_results')
+    @mock.patch('civis.io._tables._run_cleaning')
+    def test_civis_file_to_table_table_exists_all_sql_types_missing(
```
Comment (jkulzick) [inline civis/tests/test_io.py:404]: I think we want `table_does_not_exist` for this test. We're testing that if all sql types are missing that we run cleaning? But if the table exists then we don't need to run cleaning so we probably want to change this to the table not existing or the `existing_table_rows = 'drop'`.

File: civis/tests/test_io.py [hunk 4]
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
...
+        database = 'redshift-general'
+        mock_file_id = 1234
+        mock_cleaned_file_id = 1235
+        mock_import_id = 8675309
+
+        self.mock_client.imports.post_files_csv.return_value\
+            .id = mock_import_id
+        self.mock_client.get_database_id.return_value = 42
+        self.mock_client.default_credential = 713
+
```
Comment (jkulzick) [inline civis/tests/test_io.py:420]: Here we want to mock the api client response for `self.mock_client.get_table_id` based on whether or not the table exists. If the table exists, it should look like [this](https://github.com/civisanalytics/civis-python/blob/6d632e4e71888bdc63aba7655442bdc715c5f04d/civis/tests/test_io.py#L224). If the table does not exist then it should look like [this](https://github.com/civisanalytics/civis-python/blob/6d632e4e71888bdc63aba7655442bdc715c5f04d/civis/tests/test_io.py#L322).

File: civis/tests/test_io.py [hunk 5]
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
...
+        table_columns = [{'name': 'a', 'sql_type': ''},
+                         {'name': 'b', 'sql_type': ''}]
+        m_process_cleaning_results.return_value = (
+            [mock_cleaned_file_id],
+            True,  # headers
+            'gzip',  # compression
+            'comma',  # delimiter
+            table_columns  # table_columns
```
Comment (jkulzick) [inline civis/tests/test_io.py:428]: In this test, the `table_columns` provided as the parameter (to `civis.io.civis_file_to_table`) should be different than what is returned by `m_process_cleaning_results` or our cleaning job. This is because that job is detecting sql types since sql types were not provided and the table does not exist or is being dropped. So I would expect this `table_columns` to return valid `sql_type` entries.

File: civis/tests/test_io.py [hunk 6]
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
...
+            'name': expected_name,
+            'max_errors': None,
+            'existing_table_rows': 'truncate',
+            'hidden': True,
+            'column_delimiter': 'comma',
+            'compression': 'gzip',
+            'escaped': False,
+            'execution': 'immediate',
+            'loosen_types': False,
+            'table_columns': table_columns,
```
Comment (jkulzick) [inline civis/tests/test_io.py:480]: here we should expect the csv import job to be called with the `table_columns` returned by cleaning, not with the `table_columns` provided as the parameter to `civis_file_to_table`.

File: civis/tests/test_io.py [hunk 7]
```diff
@@ -398,6 +398,147 @@ def test_civis_file_to_table_table_doesnt_exist(
...
+        table_columns = [{'name': 'a', 'sql_type': 'INT'},
+                         {'name': 'b', 'sql_type': ''}]
+        m_process_cleaning_results.return_value = (
+            [mock_cleaned_file_id],
+            True,  # headers
+            'gzip',  # compression
+            'comma',  # delimiter
+            table_columns  # table_columns
+        )
+        m_run_cleaning.return_value = [mock.sentinel.cleaning_future]
```
Comment (jkulzick) [inline civis/tests/test_io.py:530]: Review what we need to mock based on what gets called before we raise the `ValueError`. For example, we shouldn't hit `post_files_csv` or `m_process_cleaning_results` yet so we shouldn't need to mock those.

We will check if the table exists so we should be mocking the api client (based on the test name the table exists). See [here](https://github.com/civisanalytics/civis-python/blob/6d632e4e71888bdc63aba7655442bdc715c5f04d/civis/tests/test_io.py#L224).

File: civis/tests/test_io.py (hunk 3 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:404]: Ah - I misread the condition statement. I was so confused.

File: civis/io/_tables.py [hunk 8]
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
     except ValueError:
         table_exists = False
 
+    table_columns_valid = True
+    if table_columns:
+        table_columns_valid = True if table_columns[0]['sql_type'] else False
+        for i in range(0, len(table_columns)-1):
+            thisSqlType = table_columns[i]['sql_type']
+            nextSqlType = table_columns[i+1]['sql_type']
+            if not thisSqlType == nextSqlType:
```
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py:1036]: Yeah this might be confusing to read, but actually putting the `not` before a string evaluates it as a boolean. So what I am checking here is that if the next sql type exists but the previous one does not, or vice versa, then we throw an error. Because that means that there is a mix between existing and non-existing sql types. If the error isn't thrown then the original assignment of True or False remains, the former means they are all provided and the latter means none are provided. I'm pretty sure this logic achieves what we wants and I like that it exits the loop as early as necessary. But let me use a test just to be sure!

File: civis/io/_tables.py [hunk 9]
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
     except ValueError:
         table_exists = False
 
+    table_columns_valid = True
+    if table_columns:
+        table_columns_valid = True if table_columns[0]['sql_type'] else False
+        for i in range(0, len(table_columns)-1):
+            thisSqlType = table_columns[i]['sql_type']
```
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py:1034]: Yes I know this, it was just out of habit haha

File: civis/io/_tables.py [hunk 10]
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
...
+                                'have a sql type provided'
+                raise ValueError(error_message)
+                break
+
     # Use Preprocess endpoint to get the table columns as needed
     # and perform necessary file cleaning
-    need_table_columns = ((not table_exists or existing_table_rows == 'drop')
+    need_table_columns = (((not table_exists or existing_table_rows == 'drop')
                           and table_columns is None)
+                          or not table_columns_valid)
```
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py:1046]: Yes I see now! I misunderstood the logic at first.

File: civis/tests/test_io.py (hunk 4 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:420]: Yeah I see now, since I had the logic statement wrong I miswrote the tests

File: civis/tests/test_io.py (hunk 5 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:428]: yeah I was wondering that! Looking at the existing tests it looked like they were passing in the same table_columns so I was confused about that, but I copied them. I'll take a closer look!

File: civis/tests/test_io.py (hunk 6 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:480]: Yeah, same confusion as above. Is it me or are these tests really hard to read...?

File: civis/tests/test_io.py (hunk 7 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:530]: Ah yes, good catch. I meant to remove these.

File: civis/io/_tables.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jkulzick) [reply civis/io/_tables.py:1036]: I think in Python we can arrive at a much simpler approach that is easier to read. Python has a lot of powerful built-ins.

For example - `sql_type_cnt = sum(1 for col in table_columns if col.get('sql_type'))`

File: civis/io/_tables.py [hunk 11]
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
...
+    if table_columns:
+        table_columns_valid = True if table_columns[0]['sql_type'] else False
+        for i in range(0, len(table_columns)-1):
+            thisSqlType = table_columns[i]['sql_type']
+            nextSqlType = table_columns[i+1]['sql_type']
+            if not thisSqlType == nextSqlType:
+                error_message = 'Some table columns do not ' \
+                                'have a sql type provided'
+                raise ValueError(error_message)
+                break
```
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py:1040]: done

File: civis/io/_tables.py [hunk 12]
```diff
@@ -1027,10 +1027,23 @@ def civis_file_to_table(file_id, database, table, client=None,
...
         table_exists = False
 
+    table_columns_valid = True
+    if table_columns:
+        table_columns_valid = True if table_columns[0]['sql_type'] else False
+        for i in range(0, len(table_columns)-1):
+            thisSqlType = table_columns[i]['sql_type']
+            nextSqlType = table_columns[i+1]['sql_type']
+            if not thisSqlType == nextSqlType:
+                error_message = 'Some table columns do not ' \
```
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py:1037]: done

File: civis/io/_tables.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py:1034]: done

File: civis/io/_tables.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py:1046]: done

File: civis/tests/test_io.py (hunk 3 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:404]: done

File: civis/tests/test_io.py (hunk 4 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:420]: done

File: civis/tests/test_io.py (hunk 5 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:428]: done

File: civis/tests/test_io.py (hunk 6 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:480]: done

File: civis/tests/test_io.py (hunk 7 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:530]: done

File: civis/io/_tables.py [hunk 13]
```diff
@@ -1040,7 +1052,7 @@ def civis_file_to_table(file_id, database, table, client=None,
         cleaning_futures, client, headers, need_table_columns, delimiter
     )
 
-    table_columns = table_columns or cleaned_table_columns
+    table_columns = cleaned_table_columns if need_table_columns else table_columns
```
Comment (izzy-zelichenko) [inline civis/io/_tables.py:1055]: This logic was flawed, since we may need to clean even when table_columns are provided, so needed to be updated.

File: civis/tests/test_io.py [hunk 14]
```diff
@@ -320,7 +320,8 @@ def test_civis_file_to_table_table_doesnt_exist(
         self.mock_client.default_credential = 713
 
         self.mock_client.get_table_id.side_effect = ValueError('no table')
-        mock_columns = [{'name': 'foo', 'sql_type': 'INTEGER'}]
+        mock_columns = [{'name': 'foo', 'sql_type': 'INTEGER'},
+                        {'name': 'bar', 'sql_type': 'VARCHAR(42)'}]
```
Comment (izzy-zelichenko) [inline civis/tests/test_io.py:324]: I added a second, different sql type to show that this test passes with current logic

File: civis/io/_tables.py [hunk 15]
```diff
@@ -1027,10 +1027,22 @@ def civis_file_to_table(file_id, database, table, client=None,
     except ValueError:
         table_exists = False
 
+    table_columns_valid = True
+    if table_columns:
+        table_columns_valid = True if table_columns[0]['sql_type'] else False
+        for i in range(0, len(table_columns)-1):
+            this_sql_type = table_columns[i]['sql_type']
+            next_sql_type = table_columns[i+1]['sql_type']
+            if (not this_sql_type) == (next_sql_type):
```
Comment (izzy-zelichenko) [inline civis/io/_tables.py:1036]: This was not actually working as expected, had to add parentheses to have them evaluate as needed.

File: civis/tests/test_io.py [hunk 16]
```diff
@@ -398,6 +398,242 @@ def test_civis_file_to_table_table_doesnt_exist(
             **expected_kwargs
         )
 
+    @pytest.mark.civis_file_to_table
+    @mock.patch('civis.io._tables._process_cleaning_results')
+    @mock.patch('civis.io._tables._run_cleaning')
+    def test_civis_file_to_table_table_doesnt_exist_valid_table_columns_provided(
```
Comment (izzy-zelichenko) [inline civis/tests/test_io.py:404]: I added a test to show that when the table does not exist, but all sql_types are provided, that even if they are different, it does not throw an error and evaluates them correctly. No cleaning happens.

**Comment by izzy-zelichenko** [comment]
We now have a test for every case when table_columns are provided

File: civis/io/_tables.py (hunk 15 above)
Comment (uttercm) [reply civis/io/_tables.py:1036]: See John's comments about simplifying this approach, this takes a bit of effort to understand what exactly we are looking for. I like the idea of using the counts of how many `sql_type` are provided.

File: civis/io/_tables.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py]: Yeah that is simpler, I'll go with this approach.

[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py]: Done. That was easy!
I haven't used Python in a really long time, so it's hard for me to pull out these tools.

File: civis/io/_tables.py [hunk 17]
```diff
@@ -1027,10 +1027,22 @@ def civis_file_to_table(file_id, database, table, client=None,
     except ValueError:
         table_exists = False
 
+    table_columns_valid = True
```
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py:1030]: got it, yeah agree!

File: civis/tests/test_io.py (hunk 5 above)
Comment (izzy-zelichenko) [reply civis/tests/test_io.py:429]: Yeah that makes sense!

File: civis/io/_tables.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (izzy-zelichenko) [reply civis/io/_tables.py]: done. hows this?

**Comment by izzy-zelichenko** [comment]
I removed the test I previously added because I realized it was nearly identical to the test `test_civis_file_to_table_table_doesnt_exist_provide_table_columns`, so I just updated that existing test to have two table_columns provided to be more thorough.

**Comment by mheilman** [comment]
@izzy-zelichenko, would you mind making a followup PR to add an item to the changelog for this?

**Comment by izzy-zelichenko** [comment]
@mheilman Sure thing.
